google-cloud-clients/google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -494,24 +494,40 @@ static String getServiceAccountProjectId(String credentialsPath) {
    */
   @SuppressWarnings("unchecked")
   public ServiceT getService() {
-    if (service == null) {
+    if (shouldRefreshService(service)) {
       service = serviceFactory.create((OptionsT) this);
     }
     return service;
   }
 
+  /**
+   * @param cachedService The currently cached service object
+   * @return true if the currently cached service object should be refreshed.
+   */
+  protected boolean shouldRefreshService(ServiceT cachedService) {
```

Comment:
Should this and `shouldRefreshRpc` be static? 

It looks like we're not using anything from the instance but we're allowing for that possibility in the future. If so, should we instead provide a protected getter for the cached service/rpc?

Reply:
The idea is to let these two methods be overridable. The base `Service` and `ServiceRpc` classes are not closeable and therefore only does a `cachedService == null` check.

The `Spanner` class (a subclass of `Service`) implements `AutoCloseable` and `SpannerOptions` therefore implements a custom check in the `shouldRefreshService()` method that calls `Spanner#isClosed()` instead of the `cachedService == null` check.